capa/features/extractors/dotnet/helpers.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+from typing import Any, Dict, Tuple, Iterator
+from itertools import chain
+
+import dnfile
+from dncil.cil.body import CilMethodBody
+from dncil.cil.error import MethodBodyFormatError
+from dncil.clr.token import Token, StringToken, InvalidToken
+from dncil.cil.body.reader import CilMethodBodyReaderBase
+
+# key indexes to dotnet metadata tables
+DOTNET_META_TABLES_BY_INDEX = {table.value: table.name for table in dnfile.enums.MetadataTables}
+
+
+class DnfileMethodBodyReader(CilMethodBodyReaderBase):
+    def __init__(self, pe: dnfile.dnPE, row: dnfile.mdtable.MethodDefRow):
+        self.pe: dnfile.dnPE = pe
+        self.offset: int = self.pe.get_offset_from_rva(row.Rva)
+
+    def read(self, n: int) -> bytes:
+        data: bytes = self.pe.get_data(self.pe.get_rva_from_offset(self.offset), n)
+        self.offset += n
+        return data
+
+    def tell(self) -> int:
+        return self.offset
+
+    def seek(self, offset: int) -> int:
+        self.offset = offset
+        return self.offset
+
+
+def generate_dotnet_token(table: int, rid: int) -> int:
+    return ((table & 0xFF) << Token.TABLE_SHIFT) | (rid & Token.RID_MASK)
+
+
+def resolve_dotnet_token(pe: dnfile.dnPE, token: Token) -> Any:
+    """map generic token to string or table row"""
+    if isinstance(token, StringToken):
+        return pe.net.user_strings.get_us(token.rid).value
```

Comment:
i think we'll need more armoring around here: its possible that the strings are not encoded correctly, so we may want to return a handle to raw data here. but lets find a test case before we try to fix it.

Reply:
see comments here: https://github.com/malwarefrank/dnfile/blob/92847841e6496453598947a74eb78fa7299ad579/src/dnfile/stream.py#L120-L158

Reply:
added additional guard rails.